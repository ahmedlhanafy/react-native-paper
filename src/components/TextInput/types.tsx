import { TextInput as NativeTextInput, Animated } from 'react-native';
import { TextInputProps } from './TextInput';

export type RenderProps = {
  ref: (a: any) => undefined;
  onChangeText: (a: string) => undefined | null | undefined;
  placeholder: string | null | undefined;
  placeholderTextColor: string | null | undefined;
  editable: boolean | null | undefined;
  selectionColor: string | null | undefined;
  onFocus: () => unknown | null | undefined;
  onBlur: () => unknown | null | undefined;
  underlineColorAndroid: string | null | undefined;
  style: any;
  multiline: boolean | null | undefined;
  numberOfLines: number | null | undefined;
  value: string | null | undefined;
};
type TextInputTypesWithoutMode = Omit<
  TextInputProps,
  'mode'
>;
export type State = {
  labeled: Animated.Value;
  error: Animated.Value;
  focused: boolean;
  placeholder: string | null | undefined;
  value: string | null | undefined;
  labelLayout: { measured: boolean; width: number; height: number };
};
export type ChildTextInputProps = {
  parentState: State;
  innerRef: (ref: NativeTextInput | null | undefined) => undefined;
  onFocus?: (args: any) => undefined;
  onBlur?: (args: any) => undefined;
  onChangeText?: (value: string) => undefined;
  onLayoutAnimatedText?: (e: any) => undefined;
} & TextInputTypesWithoutMode;
export type LabelProps = {
  placeholderStyle: any;
  placeholderOpacity: number | Animated.Value;
  baseLabelTranslateX: number;
  baseLabelTranslateY: number;
  wiggleOffsetX: number;
  labelScale: number;
  fontSize: number;
  font: any;
  topPosition: number;
  paddingOffset?: { paddingHorizontal: number } | null | undefined;
  placeholderColor: string | null | undefined;
  label?: string | null | undefined;
  hasActiveOutline: boolean | null | undefined;
  activeColor: string;
  onLayoutAnimatedText: () => unknown | null | undefined;
  error: boolean | null | undefined;
};
export type InputLabelProps = { parentState: State; labelProps: LabelProps };
