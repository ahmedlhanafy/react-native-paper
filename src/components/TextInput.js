/* @flow */

import React, { Component } from 'react';
import {
  View,
  Animated,
  TextInput as NativeTextInput,
  StyleSheet,
} from 'react-native';
import Text from './Typography/Text';
import withTheme from '../core/withTheme';
import type { Theme } from '../types/Theme';

const AnimatedText = Animated.createAnimatedComponent(Text);

type Props = {
  /**
   * If true, user won't be able to interact with the component. The default value is false
   */
  disabled?: boolean,
  /**
   * The text to use for the floating label
   */
  label?: string,
  /**
   * Placeholder for the input
   */
  placeholder?: string,
  /**
   * Callback that is called when the text input's text changes. Changed text is passed as an argument to the callback handler
   */
  onChangeText?: Function,
  /**
   * Underline color of the input
   */
  underlineColor?: string,
  /**
   * Whether the input can have multiple lines
   */
  multiline: boolean,
  /**
   * The number of lines to show in the input (Android only)
   */
  numberOfLines: number,
  /**
   * Callback that is called when the text input is focused
   */
  onFocus?: Function,
  /**
   * Callback that is called when the text input is blurred
   */
  onBlur?: Function,
  /**
   * Value of the text input
   */
  value?: string,
  /**
   * Text that gives context about a fieldâ€™s input
   */
  helperText?: string,
  /**
   * Represents error state, true represents an invalid input
   */
  hasError?: boolean,
  /**
   * Text to show for error state, visible only when hasError prop is true
   */
  errorText?: string,
  style?: any,
  theme: Theme,
};

type DefaultProps = {
  disabled: boolean,
  hasError: boolean,
};

type State = {
  focused: Animated.Value,
  placeholder: ?string,
};

/**
 * TextInputs allow users to input text.
 *
 * **Usage:**
 * ```
 * class MyComponent extends React.Component {
 *   state = {
 *     text: ''
 *   };
 *
 *   render(){
 *     return (
 *       <TextInput
 *         label='Email'
 *         value={this.state.text}
 *         onChangeText={text => this.setState({ text })}
 *       />
 *     );
 *   }
 * }
 * ```
 *
 * @extends TextInput props https://facebook.github.io/react-native/docs/textinput.html#props
 *
 */

class TextInput extends Component<DefaultProps, Props, State> {
  static defaultProps = {
    disabled: false,
    hasError: false,
  };

  constructor(props: Props) {
    super(props);

    this.state = {
      focused: new Animated.Value(0),
      placeholder: '',
    };
  }

  state: State;

  componentDidUpdate(prevProps: Props) {
    if (
      (prevProps.value !== this.props.value ||
        prevProps.placeholder !== this.props.placeholder) &&
      this.props.value === ''
    ) {
      this._setPlaceholder();
    }
  }

  componentWillUnmount() {
    clearTimeout(this._timer);
  }

  _setPlaceholder = () => {
    clearTimeout(this._timer);
    this._timer = setTimeout(
      () =>
        this.setState({
          placeholder: this.props.placeholder,
        }),
      50
    );
  };

  _removePlaceholder = () =>
    this.setState({
      placeholder: '',
    });

  _timer: any;
  _root: any;
  _setRef = (c: Object) => (this._root = c);

  _animateFocus = () => {
    Animated.timing(this.state.focused, {
      toValue: 1,
      duration: 150,
    }).start(this._setPlaceholder);
  };

  _animateBlur = () => {
    this._removePlaceholder();
    Animated.timing(this.state.focused, {
      toValue: 0,
      duration: 180,
    }).start();
  };

  _handleFocus = (...args) => {
    this._animateFocus();
    if (this.props.onFocus) {
      this.props.onFocus(...args);
    }
  };

  _handleBlur = (...args) => {
    this._animateBlur();
    if (this.props.onBlur) {
      this.props.onBlur(...args);
    }
  };

  setNativeProps(...args) {
    return this._root.setNativeProps(...args);
  }

  isFocused(...args) {
    return this._root.isFocused(...args);
  }

  clear(...args) {
    return this._root.clear(...args);
  }

  focus(...args) {
    return this._root.focus(...args);
  }

  blur(...args) {
    return this._root.blur(...args);
  }

  render() {
    const {
      value,
      disabled,
      label,
      underlineColor,
      helperText,
      hasError,
      errorText,
      style,
      theme,
      ...rest
    } = this.props;
    const { colors, fonts } = theme;
    const fontFamily = fonts.regular;
    const primaryColor = colors.primary;

    let inactiveColor = colors.disabled;
    let helperTextColor = colors.helperText;
    let inputTextColor = colors.text,
      labelColor,
      bottomLineColor;

    if (disabled) {
      inputTextColor = labelColor = bottomLineColor = inactiveColor;
    } else if (hasError) {
      helperTextColor = labelColor = bottomLineColor = inactiveColor =
        colors.error;
    } else {
      labelColor = primaryColor;
      bottomLineColor = underlineColor || primaryColor;
    }

    const labelColorAnimation = this.state.focused.interpolate({
      inputRange: [0, 1],
      outputRange: [inactiveColor, labelColor],
    });

    const translateY = value
      ? -22
      : this.state.focused.interpolate({
          inputRange: [0, 1],
          outputRange: [0, -22],
        });
    const fontSize = value
      ? 12
      : this.state.focused.interpolate({
          inputRange: [0, 1],
          outputRange: [16, 12],
        });

    const labelStyle = {
      color: labelColorAnimation,
      fontFamily,
      fontSize,
      transform: [
        {
          translateY,
        },
      ],
    };

    const bottomLineStyle = {
      backgroundColor: bottomLineColor,
      transform: [{ scaleX: this.state.focused }],
      opacity: this.state.focused.interpolate({
        inputRange: [0, 0.1, 1],
        outputRange: [0, 1, 1],
      }),
    };

    return (
      <View style={style}>
        <AnimatedText
          pointerEvents="none"
          style={[styles.placeholder, labelStyle]}
        >
          {label}
        </AnimatedText>
        <NativeTextInput
          {...rest}
          value={value}
          placeholder={this.state.placeholder}
          placeholderTextColor={colors.placeholder}
          editable={!disabled}
          ref={this._setRef}
          selectionColor={labelColor}
          onFocus={this._handleFocus}
          onBlur={this._handleBlur}
          underlineColorAndroid="transparent"
          style={[
            styles.input,
            rest.multiline && styles.multiline,
            {
              color: inputTextColor,
              fontFamily,
            },
          ]}
        />
        <View pointerEvents="none" style={styles.bottomLineContainer}>
          <View
            style={[styles.bottomLine, { backgroundColor: inactiveColor }]}
          />
          <Animated.View
            style={[styles.bottomLine, styles.focusLine, bottomLineStyle]}
          />
        </View>
        {helperText &&
          !disabled && (
            <Text style={[styles.helperText, { color: helperTextColor }]}>
              {hasError ? errorText : helperText}
            </Text>
          )}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  placeholder: {
    position: 'absolute',
    left: 0,
    top: 38,
    fontSize: 16,
  },
  input: {
    minHeight: 64,
    paddingTop: 20,
    paddingBottom: 0,
    marginTop: 8,
    marginBottom: -4,
  },
  multiline: {
    paddingTop: 30,
  },
  bottomLineContainer: {
    marginBottom: 4,
    height: StyleSheet.hairlineWidth * 4,
  },
  bottomLine: {
    position: 'absolute',
    left: 0,
    right: 0,
    top: 0,
    height: StyleSheet.hairlineWidth,
  },
  focusLine: {
    height: StyleSheet.hairlineWidth * 4,
  },
  helperText: {
    textAlign: 'left',
    paddingTop: 4,
    paddingBottom: 4,
    fontSize: 12,
  },
});

export default withTheme(TextInput);
