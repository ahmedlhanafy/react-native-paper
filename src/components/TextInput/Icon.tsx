import React from 'react';
import { View, StyleSheet } from 'react-native';
import color from 'color';
import { withTheme } from '../../core/theming';

import IconButton from '../IconButton';
import { Theme } from '../../types';

type Props = React.ComponentProps<typeof IconButton> & {
  /**
   * @optional
   */
  theme: Theme;
};

export const ICON_SIZE = 24;
const ICON_OFFSET = 12;

export function renderIcon({
  icon,
  iconTopPosition,
  side,
}: {
  icon: React.ReactNode;
  iconTopPosition: number;
  side: 'left' | 'right';
}): React.ReactNode {
  // @ts-ignore
  return React.cloneElement(icon, {
    style: {
      top: iconTopPosition,
      [side]: ICON_OFFSET,
    },
  });
}

/**
 * A component used to display an icon in the textinput.
 */
class TextInputIcon extends React.Component<Props> {
  static displayName = 'TextInput.Icon';

  render() {
    const {
      icon,
      onPress,
      style,
      color: customIconColor,
      theme,
      ...rest
    } = this.props;

    const { colors, dark } = theme;
    const textColor = colors.text;
    const iconColor =
      customIconColor ||
      (dark
        ? textColor
        : color(textColor)
            .alpha(0.54)
            .rgb()
            .string());
    const rippleColor = color(textColor)
      .alpha(0.32)
      .rgb()
      .string();

    return (
      <View style={[styles.container, style]}>
        <IconButton
          icon={icon}
          style={styles.iconButton}
          size={ICON_SIZE}
          onPress={onPress}
          rippleColor={rippleColor}
          color={iconColor}
          animated
          {...rest}
        />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    position: 'absolute',
    width: ICON_SIZE,
    height: ICON_SIZE,
    justifyContent: 'center',
    alignItems: 'center',
  },
  iconButton: {
    margin: 0,
  },
});

export default withTheme(TextInputIcon);

// @component-docs ignore-next-line
export { TextInputIcon };
