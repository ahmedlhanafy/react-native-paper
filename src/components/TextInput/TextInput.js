/* @flow */

import * as React from 'react';
import { TextInput as NativeTextInput } from 'react-native';
import { polyfill } from 'react-lifecycles-compat';

import TextInputOutlined from './TextInputOutlined';
import TextInputFlat from './TextInputFlat';
import { withTheme } from '../../core/theming';
import type { Theme } from '../../types';

type RenderProps = {
  ref: any => void,
  onChangeText: string => void,
  placeholder: ?string,
  placeholderTextColor: string,
  editable?: boolean,
  selectionColor: string,
  onFocus: () => mixed,
  onBlur: () => mixed,
  underlineColorAndroid: string,
  style: any,
  multiline?: boolean,
  numberOfLines?: number,
  value?: string,
};

type Props = React.ElementConfig<typeof NativeTextInput> & {|
  /**
   * Mode of the TextInput.
   * - `flat` - flat input with an underline.
   * - `outlined` - input with an outline.
   *
   * In `outlined` mode, the background color of the label is derived from `colors.background` in theme or the `backgroundColor` style.
   * This component render TextInputOutlined or TextInputFlat based on that props
   */
  mode?: 'flat' | 'outlined',
  /**
   * If true, user won't be able to interact with the component.
   */
  disabled?: boolean,
  /**
   * The text to use for the floating label.
   */
  label?: string,
  /**
   * Placeholder for the input.
   */
  placeholder?: string,
  /**
   * Whether to style the TextInput with error style.
   */
  error?: boolean,
  /**
   * Callback that is called when the text input's text changes. Changed text is passed as an argument to the callback handler.
   */
  onChangeText?: Function,
  /**
   * Selection color of the input
   */
  selectionColor?: string,
  /**
   * Underline color of the input.
   */
  underlineColor?: string,
  /**
   * Whether the input can have multiple lines.
   */
  multiline?: boolean,
  /**
   * The number of lines to show in the input (Android only).
   */
  numberOfLines?: number,
  /**
   * Callback that is called when the text input is focused.
   */
  onFocus?: () => mixed,
  /**
   * Callback that is called when the text input is blurred.
   */
  onBlur?: () => mixed,
  /**
   *
   * Callback to render a custom input component such as `react-native-text-input-mask`
   * instead of the default `TextInput` component from `react-native`.
   *
   * Example:
   * ```js
   * <TextInput
   *   label="Phone number"
   *   render={props =>
   *     <TextInputMask
   *       {...props}
   *       mask="+[00] [000] [000] [000]"
   *     />
   *   }
   * />
   * ```
   */
  render: (props: RenderProps) => React.Node,
  /**
   * Value of the text input.
   */
  value?: string,
  style?: any,
  /**
   * @optional
   */
  theme: Theme,
|};

type State = {};

/**
 * A component to allow users to input text.
 *
 * <div class="screenshots">
 *   <figure>
 *     <img class="medium" src="screenshots/textinput-flat.focused.png" />
 *     <figcaption>Flat (focused)</figcaption>
 *   </figure>
 *   <figure>
 *     <img class="medium" src="screenshots/textinput-flat.disabled.png" />
 *     <figcaption>Flat (disabled)</figcaption>
 *   </figure>
 *   <figure>
 *     <img class="medium" src="screenshots/textinput-outlined.focused.png" />
 *     <figcaption>Outlined (focused)</figcaption>
 *   </figure>
 *   <figure>
 *     <img class="medium" src="screenshots/textinput-outlined.disabled.png" />
 *     <figcaption>Outlined (disabled)</figcaption>
 *   </figure>
 * </div>
 *
 * ## Usage
 * ```js
 * import * as React from 'react';
 * import { TextInput } from 'react-native-paper';
 *
 * export default class MyComponent extends React.Component {
 *   state = {
 *     text: ''
 *   };
 *
 *   render(){
 *     return (
 *       <TextInput
 *         label='Email'
 *         value={this.state.text}
 *         onChangeText={text => this.setState({ text })}
 *       />
 *     );
 *   }
 * }
 * ```
 *
 * @extends TextInput props https://facebook.github.io/react-native/docs/textinput.html#props
 */

class TextInput extends React.Component<Props, State> {
  static defaultProps = {
    mode: 'flat',
    disabled: false,
    error: false,
    multiline: false,
    editable: true,
    render: props => <NativeTextInput {...props} />,
  };

  render() {
    const { mode, ...rest } = this.props;

    return mode === 'outlined' ? (
      <TextInputOutlined {...rest} />
    ) : (
      <TextInputFlat {...rest} />
    );
  }
}

polyfill(TextInput);

export default withTheme(TextInput);
