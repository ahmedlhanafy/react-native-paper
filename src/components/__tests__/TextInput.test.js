import * as React from 'react';
import { render } from 'react-native-testing-library';
import TextInput from '../TextInput/TextInput';

it('renders on TextInput', () => {
 const { toJSON } = render(<TextInput testID="input" />);
 expect( toJSON() ).toMatchSnapshot();
});

it('should be find TextInput component by testID', () => {
 const { getByTestId } = render(<TextInput testID="input" />);
 const component = getByTestId('input');
 expect(component).toBeTruthy();
});