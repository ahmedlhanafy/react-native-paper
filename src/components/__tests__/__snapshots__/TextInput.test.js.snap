// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders on TextInput 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "rgb(231, 231, 231)",
        "borderTopLeftRadius": 4,
        "borderTopRightRadius": 4,
      },
      Object {},
    ]
  }
>
  <View
    style={
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0.26)",
        "bottom": 0,
        "height": 2,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "transform": Array [
          Object {
            "scaleY": 0.5,
          },
        ],
      }
    }
  />
  <View
    style={
      Object {
        "minHeight": 64,
        "paddingBottom": 0,
        "paddingTop": 0,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      editable={true}
      multiline={false}
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      placeholderTextColor="rgba(0, 0, 0, 0.54)"
      rejectResponderTermination={true}
      selectionColor="#6200ee"
      style={
        Array [
          Object {
            "flexGrow": 1,
            "margin": 0,
            "textAlign": "left",
            "zIndex": 1,
          },
          Object {
            "paddingHorizontal": 12,
          },
          Object {
            "height": 64,
          },
          Object {
            "paddingBottom": 0,
            "paddingTop": 0,
          },
          Object {
            "color": "#000000",
            "fontFamily": "System",
            "fontSize": 16,
            "fontWeight": undefined,
            "textAlignVertical": "center",
          },
        ]
      }
      testID="input"
      underlineColorAndroid="transparent"
    />
  </View>
</View>
`;
